hardware_interface/include/hardware_interface/actuator_interface.hpp

```diff
@@ -115,33 +120,202 @@ class ActuatorInterface : public rclcpp_lifecycle::node_interfaces::LifecycleNod
+
+    std::vector<std::shared_ptr<StateInterface>> state_interfaces;
+    state_interfaces.reserve(
+      unlisted_interface_descriptions.size() + joint_state_interfaces_.size());
+
+    // add InterfaceDescriptions and create the StateInterfaces from the descriptions and add to
+    // maps.
+    for (const auto & description : unlisted_interface_descriptions)
+    {
+      auto name = description.get_name();
+      unlisted_state_interfaces_.insert(std::make_pair(name, description));
+      auto state_interface = std::make_shared<StateInterface>(description);
+      actuator_states_.insert(std::make_pair(name, state_interface));
+      unlisted_states_.push_back(state_interface);
+      state_interfaces.push_back(state_interface);
+    }
+
+    for (const auto & [name, descr] : joint_state_interfaces_)
+    {
+      auto state_interface = std::make_shared<StateInterface>(descr);
+      actuator_states_.insert(std::make_pair(name, state_interface));
+      joint_states_.push_back(state_interface);
+      state_interfaces.push_back(state_interface);
+    }
+    return state_interfaces;
+  }
 
   /// Exports all command interfaces for this hardware interface.
   /**
-   * The command interfaces have to be created and transferred according
-   * to the hardware info passed in for the configuration.
+   * Old way of exporting the CommandInterfaces. If a empty vector is returned then
+   * the on_export_command_interfaces() method is called. If a vector with CommandInterfaces is
+   * returned then the exporting of the CommandInterfaces is only done with this function and the
+   * ownership is transferred to the resource manager. The set_command(...), get_command(...), ...,
+   * can then not be used.
    *
    * Note the ownership over the state interfaces is transferred to the caller.
    *
-   * \return vector of command interfaces
+   * \return vector of state interfaces
+   */
+  [[deprecated(
+    "Replaced by vector<std::shared_ptr<CommandInterface>> on_export_command_interfaces() method. "
+    "Exporting is "
+    "handled "
+    "by the Framework.")]] virtual std::vector<CommandInterface>
+  export_command_interfaces()
+  {
+    // return empty vector by default. For backward compatibility we try calling
+    // export_command_interfaces() and only when empty vector is returned call
+    // on_export_command_interfaces()
+    return {};
+  }
+
+  /**
+   * Override this method to export custom CommandInterfaces which are not defined in the URDF file.
+   * Those interfaces will be added to the unlisted_command_interfaces_ map.
+   *
+   *  Note method name is going to be changed to export_command_interfaces() as soon as the
+   * deprecated version is removed.
+   *
+   * \return vector of descriptions to the unlisted CommandInterfaces
+   */
+  virtual std::vector<hardware_interface::InterfaceDescription>
+  export_command_interface_descriptions()
+  {
+    // return empty vector by default.
+    return {};
+  }
```

Comment:
Can we still have this when we fully move to the variants now?, as we will have the value instead of a pointer. If this is the case, maybe it's better to remove it completely or wrap the variant around a shared_ptr, so passing this would still work.